File: crossguard/patch.js
const fs = require('fs');
let content = fs.readFileSync('src/city.orig.js', 'utf8');

const getFunc = (name) => {
  const lines = content.split('\n');
  let start = -1;
  let end = -1;
  for(let i=0; i<lines.length; i++) {
    if (lines[i].startsWith(`  ${name}(`)) start = i;
    if (start !== -1 && lines[i] === '  }') {
      end = i;
      break;
    }
  }
  return lines.slice(start, end+1).join('\n');
};

const m1 = getFunc('_buildBuildingsFromModels');
const m2 = getFunc('_buildBuildingsSimple');
const m3 = getFunc('_addWindows');

let target = fs.readFileSync('src/city.js', 'utf8');

const replaceFunc = (name, replacement) => {
  const lines = target.split('\n');
  let start = -1;
  let end = -1;
  for(let i=0; i<lines.length; i++) {
    if (lines[i].startsWith(`  ${name}(`)) start = i;
    if (start !== -1 && lines[i] === '  }') {
      end = i;
      break;
    }
  }
  const oldText = lines.slice(start, end+1).join('\n');
  target = target.replace(oldText, replacement);
};

replaceFunc('_buildBuildingsFromModels', m1);
replaceFunc('_buildBuildingsSimple', m2);
replaceFunc('_addWindows', m3);

fs.writeFileSync('src/city.js', target);

File: crossguard/patch.py
import sys

def get_func(filename, func_name):
    with open(filename, 'r') as f:
        lines = f.readlines()
    
    start = -1
    end = -1
    for i, line in enumerate(lines):
        if line.startswith(f"  {func_name}("):
            start = i
        if start != -1 and line.rstrip() == "  }":
            end = i
            break
            
    return "".join(lines[start:end+1])

m1 = get_func('src/city.orig.js', '_buildBuildingsFromModels')
m2 = get_func('src/city.orig.js', '_buildBuildingsSimple')
m3 = get_func('src/city.orig.js', '_addWindows')

with open('src/city.js', 'r') as f:
    target = f.read()

def replace_func(target, func_name, replacement):
    lines = target.split('\n')
    start = -1
    end = -1
    for i, line in enumerate(lines):
        if line.startswith(f"  {func_name}("):
            start = i
        if start != -1 and line == "  }":
            end = i
            break
    
    old_text = "\n".join(lines[start:end+1])
    return target.replace(old_text, replacement)

target = replace_func(target, '_buildBuildingsFromModels', m1.strip('\n'))
target = replace_func(target, '_buildBuildingsSimple', m2.strip('\n'))
target = replace_func(target, '_addWindows', m3.strip('\n'))

with open('src/city.js', 'w') as f:
    f.write(target)
